### PR TITLE
Add experimental support for a TLS over TLS tunnel.

### PR DESCRIPTION
**Pull request by maraino**
### Description
This PR adds experimental support for TLS over (m)TLS tunnel, this can be useful in configurations like:

```
step-ca (ra-mode) <--> ghostunnel server <--> step-ca
```

If the RA is started with the environment variable `STEP_TLS_TUNNEL=ghostunnel:9000` it will connect to ghostunnel server using as client certificate the identity certificates if available. And then it will talk HTTPS over that connection.

How to test it:
1. Start CA (ca.smallstep.com):
```sh
step-ca /etc/step-ca/ca.json
```
2. Create a certificate for ghostunnel:
```sh
step ca certificate ghostunnel.smallstep.com gt.crt gt.key
```
3. Start ghostunnel server (ghostunnel.smallstep.com)
```sh
ghostunnel server --cert gt.crt --key gt.key --cacert root_ca.crt \
  --listen :9000 --target ca.smallstep.com:9000 --allow-cn ra.smallstep.com
```
4. Create the certificate for the RA:
```sh
step ca certificate ra.smallstep.com ra.crt ra.key
```
5. Create the configuration for the tunnel:
```sh
$ cat $(step path)/config/tunnel.json
{
    "type": "tTLS",
    "host": "ghostunnel.smallstep.com:9000",
    "crt": "ra.crt",
    "key": "ra.key",
    "root": "root_ca.crt"
}
```
6. Start RA (ra.smallstep.com):
```
STEP_TLS_TUNNEL=$(step path)/config/tunnel.json step-ca /etc/step-ca/ra.json
```
7. Test:
```
step ca certificate --ca-url https://ra.smallstep.com:9000 jane@doe.com jane.crt jane.key
```

**Notes**:
* Currently the RA identity certificate (if available) is read on every client connection to ghostunnel, basically every time the RA needs to sign a certificate.

### CODE REVIEW

**Comment by maraino** [comment]
This is a configuration that worked for me in nginx:
```
stream {
    log_format basic '$remote_addr [$time_local] '
            '$protocol $status $bytes_sent $bytes_received '
            '$session_time';
    access_log /tmp/acces_log basic;

    upstream backend {
        hash $remote_addr consistent;
        server ca.smallstep.com:9000;
    }

    server {
        listen       8443 ssl;
        ssl_certificate      /tmp/proxy.crt;
        ssl_certificate_key  /tmp/proxy.key;
        ssl_session_cache    shared:SSL:1m;
        ssl_session_timeout  5m;
        ssl_ciphers  HIGH:!aNULL:!MD5;
        ssl_prefer_server_ciphers  on;
        ssl_client_certificate /tmp/root_ca.crt;
        ssl_verify_client on;
        proxy_pass backend;
    }
}
```

File: ca/identity/identity.go [hunk 1]
```diff
@@ -164,8 +183,26 @@ func (i *Identity) Validate() error {
 		if err := fileExists(i.Certificate); err != nil {
 			return err
 		}
-		if err := fileExists(i.Key); err != nil {
-			return err
+		return fileExists(i.Key)
+	case TunnelTLS:
+		if i.Host == "" {
+			return errors.New("tunnel.crt cannot be empty")
```
Comment (dopey) [inline ca/identity/identity.go:189]: I don't think this error message is right

File: ca/tls.go [hunk 2]
```diff
@@ -10,13 +10,65 @@ import (
...
-	"golang.org/x/net/http2"
+	"github.com/smallstep/certificates/ca/identity"
 )
 
+// mTLSDialContext will hold the dial context function to use in
+// getDefaultTransport.
+var mTLSDialContext func() func(ctx context.Context, network, address string) (net.Conn, error)
+
+func init() {
+	// STEP_TLS_TUNNEL is an environment that can be set to do an TLS over
```
Comment (dopey) [inline ca/tls.go:26]: an environment variable

**Review by dopey — APPROVED** [review]
made a few small remarks. otherwise approved.

**Pull request merged**